src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java

```diff
@@ -48,100 +57,133 @@
+   *
+   * @return {@link Map} properties genereated based on input params
+   */
+  private static Map<String, String> generateAuthProperties(@Nullable String serviceAccount,
+                                                            String serviceAccountType, String... prefixes) {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(SERVICE_ACCOUNT_TYPE, serviceAccountType);
+    if (serviceAccount != null) {
+      properties.put(SERVICE_ACCOUNT, serviceAccount);
+    }
 
-      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_EMAIL_SUFFIX), credentials.getClientEmail());
-      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_PRIVATE_KEY_ID_SUFFIX),
-                     credentials.getPrivateKeyId());
-      if (privateKeyData == null) {
-        privateKeyData = extractPrivateKey(credentials);
-      }
-      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_KEY_SUFFIX), privateKeyData);
-      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_JSON_SUFFIX), serviceAccount);
+    // use a custom AccessTokenProvider that uses the newer
+    // AccessTokenProviderClassFromConfigFactory will by default look for
+    //   google.cloud.auth.access.token.provider.impl
+    // but can be configured to also look for the conf with other prefixes like
+    //   gs.fs.auth.access.token.provider.impl
+    //   mapred.bq.auth.access.token.provider.impl
+    // for use by GCS and BQ.
+    for (String prefix : prefixes) {
+      properties.put(prefix + AccessTokenProviderClassFromConfigFactory.ACCESS_TOKEN_PROVIDER_IMPL_SUFFIX,
+                     ServiceAccountAccessTokenProvider.class.getName());
     }
     return properties;
   }
 
-  public static Map<String, String> getFileSystemProperties(GCPConfig config, String path,
-                                                            Map<String, String> properties) {
-    return getFileSystemProperties(config.getProject(), config.getServiceAccount(), config.getServiceAccountType(),
-                                   path, properties);
+  /**
+   * Loads credentials based on configuration properties set in the conf. This assumes that the properties from
+   * {@link #generateAuthProperties(String, String, String...)} were placed in the configuration.
+   */
+  public static GoogleCredentials loadCredentialsFromConf(Configuration conf) throws IOException {
+    String serviceAccountType = conf.get(GCPUtils.SERVICE_ACCOUNT_TYPE);
+    String serviceAccount = conf.get(GCPUtils.SERVICE_ACCOUNT);
+
+    GoogleCredentials credentials;
+    if (serviceAccount == null) {
+      credentials = GoogleCredentials.getApplicationDefault().createScoped();
+    } else {
+      boolean isFile = SERVICE_ACCOUNT_TYPE_FILE_PATH.equals(serviceAccountType);
+      credentials = loadServiceAccountCredentials(serviceAccount, isFile);
+    }
+
+    return credentials.createScoped(CredentialFactory.GCS_SCOPES);
```

Comment:
Will this only ever need GCS_SCOPES? If so, can we rename it to be `loadGCSCredentialsFromConf`? Just want to make sure this method doesn't get mistakenly called for some other client. Thanks!

Reply:
good catch, it's not always supposed to be. Will introduce another version of the method that takes the scopes. If none are given, it won't scope the credential.